src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+            }
+
+            if (!this.PendingBatch.Keys.OrderBy(s => s).SequenceEqual(this.PendingBatch.Keys))
+            {
+                this.logger.LogTrace("(-)[PENDING_BATCH_INCORRECT_SEQEUNCE]");
+                throw new InvalidOperationException("Invalid ProvenBlockHeader pending batch sequence - unable to save to the database repository.");
+            }
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                var pendingBatchKeys = this.PendingBatch.Keys.ToList();
+
+                List<ProvenBlockHeader> pendingBatchValues = this.PendingBatch.Values.ToList();
+
+                await this.provenBlockHeaderRepository.PutAsync(
+                    pendingBatchValues, this.pendingTipHashHeight).ConfigureAwait(false);
+
+                pendingBatchKeys.ForEach(
+                    blockHeight => this.PendingBatch.TryRemove(blockHeight, out ProvenBlockHeader unused));
+
+                this.pendingTipHashHeight = null;
+            }
+        }
+
+        /// <summary>
+        /// Gets <see cref="ProvenBlockHeader"></see> items from cache with specific key if present.
+        /// </summary>
+        /// <param name="key"> Item's key.</param>
+        /// <param name = "items"> Dictionary of items in cache</param>
+        /// <returns> <see cref="ProvenBlockHeader"></see> if cache contains the item, <c>null</c> otherwise.</returns>
+        private ProvenBlockHeader GetHeaderFromCache(int key, ConcurrentDictionary<int, ProvenBlockHeader> items)
+        {
+            if (items.TryGetValue(key, out ProvenBlockHeader header))
+            {
+                this.logger.LogTrace("(-):{0}", header);
+
+                return header;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Manages store cache size. Remove 30% of items if the memory cache size has been reached.
+        /// </summary>
+        private void ManangeCacheSize()
+        {
+            do
+            {
+                if (this.CacheSizeInBytes > this.MemoryCacheSizeLimitInBytes)
+                {
+                    List<int> sortedList = this.Cache.Keys.OrderBy(s => s).ToList();
+
+                    var itemsToRemove = sortedList.Take(Convert.ToInt32(sortedList.Count * 0.3));
+
+                    foreach (var itemToRemove in itemsToRemove)
+                    {
+                        if (this.Cache.TryRemove(itemToRemove, out ProvenBlockHeader header))
+                            this.CacheSizeInBytes -= header.HeaderSize;
+                    }
+                }
+
+            } while (this.CacheSizeInBytes > this.MemoryCacheSizeLimitInBytes);
+        }
+
+        /// <summary>
+        /// Check pending cache first. Then check store cache next.
+        /// </summary>
+        /// <param name="blockHeight">Item's key.</param>
+        /// <returns><see cref="ProvenBlockHeader"> if found, otherwise null.</returns>
```

Comment:
not closed xml element

Reply:
ok, checked the rest of the code files as well.

Reply:
CheckCache is confusing maybe better is CheckCacheAndPending 
if we push pending to cache though then we dont need to check pending.